File: crap4ts/scripts/lib/analyzer.mjs
import { coverageForRange } from "./coverage.mjs";

export const moduleName = (filePath) => filePath.replace(/\.[cm]?[jt]sx?$/, "");

const nestedRanges = (target, functions) =>
  functions
    .filter(
      (candidate) =>
        candidate !== target &&
        candidate.file === target.file &&
        candidate.startLine >= target.startLine &&
        candidate.endLine <= target.endLine,
    )
    .map((candidate) => [candidate.startLine, candidate.endLine]);

export function buildEntries(functions, coverageByFile) {
  return functions.map((current) => {
    const statements = coverageByFile.get(current.file);
    const measured =
      statements === undefined
        ? null
        : coverageForRange(
            statements,
            current.startLine,
            current.endLine,
            nestedRanges(current, functions),
          );
    return {
      name: current.name,
      module: moduleName(current.file),
      file: current.file,
      startLine: current.startLine,
      endLine: current.endLine,
      complexity: current.complexity,
      coverage: measured === null ? null : measured.coverage,
      statements: measured === null ? 0 : measured.statements,
      coveredStatements: measured === null ? 0 : measured.covered,
    };
  });
}

File: crap4py/scripts/crapkit/analyzer.py
from crapkit.complexity import Block
from crapkit.coverage_data import FunctionCoverage, normalize_path
from crapkit.score import Entry


def module_name(path: str) -> str:
    return normalize_path(path).removesuffix(".py").replace("/", ".")


def build_entries(
    blocks: list[Block],
    coverage: dict[str, dict[int, FunctionCoverage]],
) -> list[Entry]:
    entries: list[Entry] = []
    for block in blocks:
        path = normalize_path(block.file)
        functions = coverage.get(path)
        measured = None if functions is None else functions.get(block.start_line)
        entries.append(
            Entry(
                name=block.name,
                module=module_name(path),
                file=path,
                start_line=block.start_line,
                end_line=block.end_line,
                complexity=block.complexity,
                coverage=_coverage_of(functions, measured),
                statements=measured.statements if measured else 0,
                covered_statements=measured.covered if measured else 0,
            )
        )
    return entries


def _coverage_of(
    functions: dict[int, FunctionCoverage] | None,
    measured: FunctionCoverage | None,
) -> float | None:
    if functions is None:
        return None
    if measured is None:
        return 0.0
    return measured.percent
